cli/src/main.rs

```diff
@@ -341,11 +431,246 @@ impl CliConfig for OxideOverride {
+            unimplemented!("nested array of items")
+        }
+    };
+
+    get_referenced_type_fields(root_schema, &item_type)
+}
+
+/// Get the field names of a type using its name.
+fn get_referenced_type_fields(root_schema: &RootSchema, type_name: &str) -> Vec<String> {
+    let item_schema = root_schema
+        .definitions
+        .get(type_name)
+        .unwrap()
+        .clone()
+        .into_object();
+
+    // Generally child items are a single type, but a few APIs return an enum, e.g.
+    // `/v1/anti-affinity-groups/{anti_affinity_group}/members`. In this case we collect all of the
+    // fields present on the variants.
+    match (
+        item_schema.object.as_ref(),
+        item_schema.subschemas.and_then(|s| s.one_of),
+    ) {
+        (Some(obj), _) => obj.properties.keys().cloned().collect(),
+        (_, Some(variants)) => collect_variant_fields(&variants),
+        _ => unimplemented!("item was neither an object nor one_of subschema"),
+    }
+}
+
+/// Gather the fields present on each variant, removing duplicates and retaining their original order.
+fn collect_variant_fields(variants: &[Schema]) -> Vec<String> {
+    let mut fields = IndexSet::new();
+    for variant in variants {
+        fields.extend(
+            variant
+                .clone()
+                .into_object()
+                .object
+                .as_ref()
+                .expect("variant was not an object")
+                .properties
+                .keys()
+                .cloned(),
+        );
+    }
+    fields.into_iter().collect()
+}
+
+/// Set a table's header to the fields available to be printed.
+fn set_header_fields(
+    requested_fields: &[String],
+    schema_fields: Vec<String>,
+    table: &mut Table,
+) -> Vec<String> {
+    let printable_fields = if !requested_fields.is_empty() {
+        let requested: HashSet<_> = requested_fields.iter().collect();
+        let available: HashSet<_> = schema_fields.iter().collect();
+        let invalid = requested.difference(&available);
+
+        for field in invalid {
+            eprintln_nopipe!("WARNING: '{field}' is not a valid field");
+        }
+
+        let mut fields = requested_fields.to_vec();
+        fields.retain(|f| available.contains(f));
+        fields
+    } else {
+        schema_fields
+    };
+
+    let upcased: Vec<_> = printable_fields.iter().map(|f| f.to_uppercase()).collect();
+
+    table.set_header(upcased);
+    printable_fields
+}
+
+/// Format an object's fields for printing in a table.
+fn create_row(fields: &[String], obj: &serde_json::Map<String, serde_json::Value>) -> Vec<String> {
+    let mut row = Vec::with_capacity(fields.len());
+
+    for field in fields {
+        let s = obj.get(field).map(|f| f.to_string()).unwrap_or_default();
+
+        // `to_string` encloses values in double quotes.
+        if s.contains(' ') {
+            row.push(s);
+        } else {
+            row.push(s.trim_matches('"').to_string());
+        }
```

Comment:
Is there a test exercising this space logic? Might be good to comment about why it is this way.

Reply:
Good call, added [a test](https://github.com/oxidecomputer/oxide.rs/blob/d005c780dc7bf8eb1d543ba9733a5389f2b82b87/cli/tests/test_table.rs#L14-L56) and [commented](https://github.com/oxidecomputer/oxide.rs/blob/d005c780dc7bf8eb1d543ba9733a5389f2b82b87/cli/src/cli_builder.rs#L48-L49).